File: Internet-services-architecture-and-development/Esercitazione4/Soluzioni/verifica_disponibilità_vini_20231206/client-naive.c
#define _POSIX_C_SOURCE 200809L
#include <stdio.h>
#include <stdlib.h>
#include <sys/wait.h>
#include <fcntl.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netdb.h>

#define DIM 4096

int main(int argc, char** argv) {
        int err;
        struct addrinfo hints;
        struct addrinfo *res,*ptr;
        char *host_remoto;
        char *servizio_remoto;
        int sd, nread;
        char buf[DIM], nome[DIM], annata[DIM];
        int i=1;
        char *ackVer="ack\n";

        /* Controllo argomenti */
        if (argc < 3) {
                printf("Uso: verifica_disponibilità_vini <server> <porta>\n");
                exit(EXIT_FAILURE);
        }

        /* Costruzione dell'indirizzo */
        memset(&hints, 0, sizeof(hints));
        hints.ai_family   = AF_INET;
        hints.ai_socktype = SOCK_STREAM;

        /* Risoluzione dell'host */
        host_remoto = argv[1];
        servizio_remoto = argv[2];

        if ((err = getaddrinfo(host_remoto, servizio_remoto, &hints, &res)) != 0) {
                fprintf(stderr, "Errore risoluzione nome: %s\n", gai_strerror(err));
                exit(EXIT_FAILURE);
        }

        for (;;) {
                /* Lettura dati dall'utente */
                printf("Inserisci il nome di un vino cui sei interessato ('fine' per uscire):\n");
                if(scanf("%s", nome) == EOF){ //ho il terminatore
                        perror("scanf");
                        exit(EXIT_FAILURE);
                } 

                if (strcmp(nome, "fine") == 0) {
                        printf("Hai scelto di terminare il programma.\n");
                        break;
                }

                /* Lettura dati dall'utente */
                printf("Inserisci l'annata di un vino cui sei interessato ('fine' per uscire):\n");
                if(scanf("%s", annata) == EOF){ //ho il terminatore
                        perror("scanf");
                        exit(EXIT_FAILURE);
                }

                if (strcmp(annata, "fine") == 0) {
                        printf("Hai scelto di terminare il programma.\n");
                        break;
                }

                /* Connessione con fallback */
                for (ptr = res; ptr != NULL; ptr = ptr->ai_next) {
                        /*se socket fallisce salto direttamente alla prossima iterazione*/
                        if ((sd = socket(ptr->ai_family, ptr->ai_socktype, ptr->ai_protocol)) < 0) {
                                fprintf(stderr,"creazione socket fallita\n");
                                continue;
                        }
                        /*se connect funziona esco dal ciclo*/
                        if (connect(sd, ptr->ai_addr, ptr->ai_addrlen) == 0) {
                                printf("connect riuscita al tentativo %d\n",i);
                                break;
                        }
                        i++;
                        close(sd);
                }

                /* Verifica sul risultato restituito da getaddrinfo */
                if (ptr == NULL) {
                        fprintf(stderr, "Errore risoluzione nome: nessun indirizzo corrispondente trovato\n");
                        exit(EXIT_FAILURE);
                }

                /* Mando il nome del vino al server */
                if (write(sd, nome, strlen(nome)) < 0) {
                        perror("write nome");
                        exit(EXIT_FAILURE);
                }

                /* Leggo l'ack */
                memset(buf, 0, sizeof(buf));
                if (read(sd, buf, sizeof(buf)-1) < 0) {
                        perror("read ack");
                        exit(EXIT_FAILURE);
                }

                if (strcmp(buf, ackVer) != 0) {
                        close(sd);
                        printf("errore nell'ack\n");
                        exit(EXIT_FAILURE);
                }

                /* Mando l'annata del vino al server */
                if (write(sd, annata, strlen(annata)) < 0) {
                        perror("write annata");
                        exit(EXIT_FAILURE);
                }

                /* Ricezione risultato */

                /* È consigliabile effettuare il flushing del buffer di printf
                 * prima di iniziare a scrivere sullo standard output con write */
                fflush(stdout);

                while ((nread = read(sd, buf, DIM)) > 0) {
                        if (write(1, buf, nread) < 0) {
                                perror("write su stdout");
                                exit(EXIT_FAILURE);
                        }
                }

                /* Controllo errori di lettura */
                if (nread < 0) {
                        perror("read del risultato");
                        exit(EXIT_FAILURE);
                }

                /* Chiudo la socket */
                close(sd);
        }

        /* Liberiamo la memoria allocata da getaddrinfo() */
        freeaddrinfo(res);

        return 0;
}

File: III - Architettura e sviluppo servizi internet/Esercizi/Esercitazione4/Soluzioni/verifica_disponibilità_vini_20231206/ClientNaive.java
import java.io.*;
import java.net.*;

public class ClientNaive {
    public static void main(String[] argv) {

        if (argv.length != 2) {
            System.err.println("Uso corretto: java ClientNaive server porta");
            System.exit(1);
        }

        try{
            BufferedReader fromUser = new BufferedReader(new InputStreamReader(System.in));

            while (true) {
                System.out.println("Inserisci il nome di un vino cui sei interessato ('fine' per uscire): ");
                String nomeVino = fromUser.readLine();

                if (nomeVino.equals("fine")){
                    System.out.println("Hai scelto di terminare il programma.");
                    break;
                }

                System.out.println("Inserisci l'annata di un vino cui sei interessato ('fine' per uscire): ");
                String annata = fromUser.readLine();

                if (annata.equals("fine")){
                    System.out.println("Hai scelto di terminare il programma.");
                    break;
                }

                Socket s = new Socket(argv[0], Integer.parseInt(argv[1]));

                BufferedReader fromServer = new BufferedReader(new InputStreamReader(s.getInputStream(),"UTF-8"));
                BufferedWriter toServer = new BufferedWriter(new OutputStreamWriter(s.getOutputStream(),"UTF-8"));

                toServer.write(nomeVino);
                toServer.flush();

                String buff = fromServer.readLine();

                if(!buff.equals("ack")){
                    System.err.println("Errore lettura Ack dal server");
                    System.exit(2);
                }

                toServer.write(annata);
                toServer.flush();

                while ((buff = fromServer.readLine()) != null) {
                    System.out.println(buff);
                }

                s.close();
            }

        }
        catch(IOException e){
            System.err.println(e.getMessage());
            e.printStackTrace();
            System.exit(3);
        }

    }
}
